zap/src/irgen/des.rs

```diff
@@ -291,6 +291,75 @@ impl Des<'_> {
 				self.push_struct(struct_ty, into)
 			}
 
+			Ty::Or(or_tys, discriminant_numty) => {
+				let (into_ty_i_name, into_ty_i_expr) = self.add_occurrence("ty_i");
+
+				self.push_local(into_ty_i_name, Some(self.readnumty(*discriminant_numty)));
+				let mut initial_if = true;
+				let mut i_offset = 0usize;
+
+				for (i, ty) in or_tys.iter().enumerate() {
+					let i = i + i_offset;
+
+					match ty.primitive_ty() {
+						PrimitiveTy::Enum(Enum::Unit(variants)) => {
+							i_offset += variants.len() - 1;
+
+							for (offset, variant) in variants.into_iter().enumerate() {
+								let condition = into_ty_i_expr.clone().eq(((i + offset) as f64).into());
+								if initial_if {
+									self.push_stmt(Stmt::If(condition));
+									initial_if = false;
+								} else {
+									self.push_stmt(Stmt::ElseIf(condition));
+								}
+
+								self.push_assign(into.clone(), Expr::Str(variant.to_string()));
+							}
+
+							continue;
+						}
+						PrimitiveTy::Enum(Enum::Tagged { tag, variants }) => {
+							i_offset += variants.len() - 1;
+
+							for (offset, (variant, data)) in variants.into_iter().enumerate() {
+								let condition = into_ty_i_expr.clone().eq(((i + offset) as f64).into());
+								if initial_if {
+									self.push_stmt(Stmt::If(condition));
+									initial_if = false;
+								} else {
+									self.push_stmt(Stmt::ElseIf(condition));
+								}
+
+								self.push_assign(into.clone(), Expr::EmptyTable);
+								self.push_assign(
+									into.clone().eindex(Expr::Str(tag.to_string())),
+									Expr::Str(variant.to_string()),
+								);
+								self.push_struct(&data, into.clone());
+							}
+
+							continue;
+						}
+						_ => {}
```

Comment:
Instead of shadowing on line 302, maybe we do `i_offset += 1` here, and remove the `enumerate` on line 301.